src/mutation_data.cpp

```diff
@@ -387,6 +399,209 @@ void mutation_branch::reset_all()
+            if (traits.test_string()) {
+                tg->add_trait_entry(trait_id(traits.next_string()), 100);
+            } else if (traits.test_array()) {
+                JsonArray subtrait = traits.next_array();
+                tg->add_trait_entry(trait_id(subtrait.get_string(0)), subtrait.get_int(1));
+            } else {
+                JsonObject subobj = traits.next_object();
+                add_entry(tg, subobj);
+            }
+        }
+    }
+    if (jsobj.has_member("groups")) {
+        JsonArray traits = jsobj.get_array("groups");
+        while (traits.has_more()) {
+            if (traits.test_string()) {
+                tg->add_group_entry(traits.next_string(), 100);
+            } else if (traits.test_array()) {
+                JsonArray subtrait = traits.next_array();
+                tg->add_group_entry(subtrait.get_string(0), subtrait.get_int(1));
+            } else {
+                JsonObject subobj = traits.next_object();
+                add_entry(tg, subobj);
+            }
+        }
+    }
+}
+
+void mutation_branch::add_entry(Trait_group *tg, JsonObject &obj) {
+    std::unique_ptr<Trait_creation_data> ptr;
+    int probability = obj.get_int("prob", 100);
+    JsonArray jarr;
+
+    if (obj.has_member("collection")) {
+        ptr.reset(new Trait_group_collection(probability));
+        jarr = obj.get_array("collection");
+    } else if (obj.has_member("distribution")) {
+        ptr.reset(new Trait_group_distribution(probability));
+        jarr = obj.get_array("distribution");
+    }
+
+    if (ptr) {
+        Trait_group *tg2 = dynamic_cast<Trait_group *>(ptr.get());
+        while (jarr.has_more()) {
+            JsonObject job2 = jarr.next_object();
+            add_entry(tg2, job2);
+        }
+        tg->add_entry(ptr);
+        return;
+    }
+
+    if (obj.has_member("trait")) {
+        trait_id id(obj.get_string("trait"));
+        ptr.reset(new Single_trait_creator(id, probability));
+    } else if (obj.has_member("group")) {
+        ptr.reset(new Trait_group_creator(obj.get_string("group"), probability));
+    }
+
+    if (!ptr) {
+        return;
+    }
+
+    tg->add_entry(ptr);
+}
+
+void mutation_branch::finalize() {
+    finalize_trait_blacklist();
+}
+
+void mutation_branch::finalize_trait_blacklist() {
+    for (auto &trait : trait_blacklist) {
+        if (!has_trait(trait)) {
+            debugmsg("trait on blacklist %s does not exist", trait.c_str());
+        }
+    }
+
+    for (auto &md: mutation_data) {
+        if (trait_is_blacklisted(md.first)) {
+            for (auto &grp : trait_groups) {
+                grp.second->remove_trait(md.first);
+            }
+        }
+    }
+}
+
+Trait_creation_data* mutation_branch::get_group( const Trait_group_tag &gid ) {
+    if (trait_groups.count(gid) > 0) {
+        return trait_groups[gid];
+    }
+    return nullptr;
+}
+
+std::vector<Trait_group_tag> mutation_branch::get_all_group_names() {
+    std::vector<std::string> rval;
```

Comment:
Gotcha. Why are you returning a `vector<string>` while the function declaration says it's returning a `vector<Trait_group_tag>`? This would not happen with `string_id`.